extensions/open-with-app/src/index.tsx

```diff
@@ -1,70 +1,188 @@
+        set end of selectedPaths to selectedItemPath
+      end repeat
+     
+      return selectedPaths as list
+    end tell
+  `)) as string;
+  if (paths.length === 0) return [];
+  return paths.split(", ").map((path) => ({ path }));
+}
 
-  useEffect(() => {
-    async function fetchApplications() {
-      setApplications(await getApplications());
+/**
+ * Get the extensions of the selected items.
+ * Files without extensions are grouped under "file".
+ */
+function getExtensions(items: FileSystemItem[]) {
+  const extensions = new Set<string>();
+  for (const item of items) {
+    const parts = item.path.split("/").at(-1)?.split(".");
+    if (!parts || parts.length < 2) {
+      extensions.add("file");
+    } else {
+      extensions.add(parts.at(-1)!);
     }
+  }
+  return Array.from(extensions).sort();
+}
 
-    fetchApplications();
-  }, []);
-
-  return (
-    <List isLoading={applications.length === 0} searchBarPlaceholder="Filter applications by name...">
-      {applications.map((application) => (
-        <ApplicationsListItem key={application.bundleId} application={application} />
-      ))}
-    </List>
-  );
+const home = homedir();
+/**
+ * Trim the path to make it more readable.
+ */
+function trimPath(path: string) {
+  if (path.startsWith(home)) {
+    path = "~" + path.slice(home.length);
+  }
+  const parts = path.split("/");
+  if (parts.length > 4) {
+    path = parts[0] + "/" + parts[1] + "/…/" + parts.at(-2) + "/" + parts.at(-1);
+  }
+  return path;
 }
 
-function ApplicationsListItem(props: { application: Application }) {
-  const application = props.application;
+/**
+ * Generate descriptive title that fits all the selected items.
+ */
+function makeTitle(items: FileSystemItem[]) {
+  // empty selection
+  if (items.length === 0) return "Open with…";
+  // single item
+  if (items.length === 1) return `Open ${trimPath(items[0].path)} with…`;
 
+  const parent = items[0].path.split("/").slice(0, -1).join("/");
+  const canUseParent = parent && items.every((item) => item.path.startsWith(parent));
+
+  // all items have same extension
+  const extensions = getExtensions(items);
+  if (extensions.length === 1 && extensions[0] !== "file") {
+    return `Open ${items.length} ${extensions[0]} files${canUseParent ? ` from ${trimPath(parent)}` : ""} with…`;
+  }
+  // all items have same parent directory
+  if (canUseParent) {
+    return `Open ${items.length} items from ${trimPath(parent)} with…`;
+  }
+  // fallback
+  return `Open ${items.length} items with…`;
+}
+
+function ApplicationListItem({
+  app,
+  items,
+  onAction,
+}: {
+  app: Application;
+  items: FileSystemItem[];
+  onAction?: (app: Application) => void;
+}) {
+  async function openAction() {
+    await closeMainWindow({ clearRootSearch: true, popToRootType: PopToRootType.Suspended });
+    onAction?.(app);
+    await Promise.all(items.map((item) => open(item.path, app)));
+    await popToRoot({ clearSearchBar: true });
+  }
```

Comment:
@fturcheti do you remember this?

Reply:
"I think we used `popToRoot` after `closeMainWindow` because we wanted that behavior, and `closeMainWindow` didn't have the `options` parameter it has nowadays.

Personally, I prefer these kinds of commands to pop to the root (because it's unlikely that you will open the files twice). But what do you recommend, @pernielsentikaer? Is it better practice to respect the user setting for "pop to root" even in this situation?"

Reply:
Lets keep it as is for now and get back to it if needed 🙂